Fixes 'Can't update during existing transition' warning message

crob611 opened the pull request:
## Summary

This gets rid of that "React cannot update state during an existing state transition" warning that is showing up in the console in canvas.  

So what was happening, to the best I can understand, is that `refresh_controls` component has this bit of code 
```js
if (inFlight || pending > 0) {
  loadingIndicator.show();
} else {
  loadingIndicator.hide();
}
```
so it will either increment or decrement the value which causes an observable value change.  `loading_indicator.js` is subscribed to that observable, and is setting it's state based on the stream it receives.  

Because `refresh_controls` is an SFC, this is all happening during a render, so we are doing a `setState` during a render, hence the warning.  

The fix, which is maybe a little hacky, is to delay actually changing the observable until after the render cycle is completed by wrapping it into a setTimeout.  If this was entirely a React environment, they would probably be moved to an action of some sort or something, but it's not, so _shrug_  

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- ~[ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~
- ~[ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~
- ~[ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~
- ~[ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios~
- ~[ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
Pinging @elastic/kibana-canvas

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6867/)

w33ble commented on x-pack/plugins/canvas/public/lib/loading_indicator.ts:
```diff
@@ -12,13 +12,15 @@ let isActive = false;
-      loadingCount.increment();
+      // Using setTimeout ensures the loadingCount observables don't change in the middle
+      // of a render and cause the "can't set state during existing transition" warning
+      setTimeout(loadingCount.increment, 0);
```
Nice job chasing that down! I don't think the loading indicator should be "async" like this though, especially if you're not exposing a way to check when the operation completes.

This is really a bug in `RefreshControl`, it shouldn't be setting that value at render time. `setTimeout` is a quick and dirty fix, but not really the right one (as you note), but if you're going to use it, use it in that component instead:

```js
if (inFlight || pending > 0) {
    setTimeout(loadingIndicator.show, 0);
  } else {
    setTimeout(loadingIndicator.hide, 0);
  }
```

Or wrap that up in a new function prop that fakes the async stuff internally:

```js
// container
withProp(props => ({
  showLoadingIndicator: show => setTimeout(() => show ? loadingIndicator.show() : loadingIndicator.hide(), 0),
}));

// component
if (inFlight || pending > 0) {
    showLoadingIndicator(true);
  } else {
    showLoadingIndicator(false);
  }
```

Ideally though, this would happen before the component even renders. You could possibly do this in the container, but the right place is probably somewhere in redux as a side-effect, either in the reducer or in some middleware.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7825/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7831/)

crob611 commented:
@w33ble This is ready for another look.  Moved all the logic into some middleware, and went a little further and converted the middleware over to typescript and added tests for it.  Tried to break up the commits in such a way that it would be easiest to walk through them 1 by 1 vs looking at the overall changes.

w33ble commented on x-pack/plugins/canvas/public/lib/loading_indicator.ts:
```diff
@@ -9,6 +9,11 @@ import { loadingCount } from 'ui/chrome';
 
 let isActive = false;
 
+export interface LoadingIndicatorInterface {
```
TS question: Shouldn't you tie this interface to the `loadingIndicator` object somehow?

crob611 replied:
Typescript is able to infer that loadingIndicator conforms to this interface.

So, when we create the middleware, we have defined it should be given an argument called `loadingIndicator` that matches the shape defined in this interface.  The compiler is able to look at this `loadingIndicator`, or the dummy one that we make in the test file, and see that they conform.  

_I think_

w33ble replied:
Yeah, I know it can infer types, which is why you normally don't need to provide a return type for functions. I thought maybe that also meant that you could import `loadingIndicator` as a type without the extra Interface definition, but I don't know. What you have it fine, just trying to learn.

> I think

😆
